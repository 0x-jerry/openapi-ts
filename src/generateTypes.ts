import type { SchemaObject } from 'openapi-typescript'
import { compile } from 'json-schema-to-typescript'

export async function generateInterface(name: string, schema: SchemaObject) {
  const result = await compile(schema as any, name, {
    bannerComment: '',
    format: false,
    unknownAny: true,
    unreachableDefinitions: false,
  })

  return result
}
