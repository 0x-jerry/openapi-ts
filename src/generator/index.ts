import type { GeneratorContext } from './core'
import { generateFlattenStyle } from './generateFlattenStyle'
import { generateNestedStyle } from './generateNestedStyle'

export interface GenerateOption {
  style: 'nested' | 'flatten'
}

export async function generateFromCtx(ctx: GeneratorContext, opt: GenerateOption) {
  if (opt.style === 'flatten') {
    return generateFlattenStyle(ctx)
  }

  return generateNestedStyle(ctx)
}

export type { GeneratorContext }
