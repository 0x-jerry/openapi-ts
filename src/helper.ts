import type { OpenAPI3, ReferenceObject, SchemaObject } from 'openapi-typescript'
import { type Arrayable, ensureArray } from '@0x-jerry/utils'
import type { ParserContext } from './parser'

export interface ReplaceSchemaTypeConfig {
  type: string
  format?: string[]

  targetType: string
  targetFormat?: string
}

/**
 * Replace some type in the schema
 *
 * @example
 * transform number type to string type if format is 'int64'
 *
 * ```ts
 * fixRawType(ctx, [{
 *  type: 'number',
 *  format: ['int64'],
 *  targetType: 'string'
 * }])
 * ```
 *
 * @param ctx
 * @param opt
 */
export function replaceSchemaType(ctx: ParserContext, opt: Arrayable<ReplaceSchemaTypeConfig>) {
  const schemas: SchemaObject[] = Object.values(ctx.schema.components?.schemas || {})

  const conf = ensureArray(opt)

  // fix endless loop
  const visited = new Set<SchemaObject>()

  schemas.forEach((schema) => fixType(schema))

  function fixType(schema: SchemaObject) {
    if (visited.has(schema)) {
      return
    }

    visited.add(schema)

    if (!('type' in schema)) {
      return
    }

    if (schema.type === 'object') {
      for (const key in schema.properties) {
        fixType(getRef(ctx, schema.properties[key]))
      }
    } else if (schema.type === 'array') {
      if (schema.items) {
        const items = ensureArray(schema.items)
        for (const item of items) {
          fixType(getRef(ctx, item))
        }
      }
    } else {
      for (const item of conf) {
        const isTheSameType = schema.type === item.type
        const hasFormat = item.format ? item.format.includes(schema.format!) : true

        if (isTheSameType && hasFormat) {
          // @ts-ignore
          schema.type = item.targetType
          schema.format = item.targetFormat
        }

        console.log(isTheSameType, hasFormat, schema)
      }
    }
  }
}

export type DereferenceObject<T> = [T] extends [infer U | ReferenceObject] ? U : T

export function getRef<T>(
  ctx: {
    schema: OpenAPI3
  },
  type: T,
): DereferenceObject<T> {
  if (!isRef(type)) {
    return type as any
  }

  let r: any = ctx.schema

  const paths = type.$ref.split('/').slice(1)

  paths.forEach((key) => (r = r[key]))

  return getRef(ctx, r)
}

export function isRef(o: any): o is ReferenceObject {
  return !!o?.$ref
}
