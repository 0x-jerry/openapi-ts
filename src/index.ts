export * from './parser'
export * from './swagger2openapi'
export * from './generator'
export * from './helper'
