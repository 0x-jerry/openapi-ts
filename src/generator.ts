import { groupBy } from 'lodash-es'
import type { ParserContext, APIConfig, APIParameterConfig } from './parser'
import { type IFs } from 'memfs'
import path, { dirname } from 'path'
import type { SchemaObject } from 'openapi-typescript'
import { generateInterface } from './generateTypes'
import { createSpinner } from './utils'
import type { Volume } from 'memfs/lib/volume'
import { camelCase } from '@0x-jerry/utils'

const config = {
  nameMapper: {
    requestAdapterName: '_request',
    apiModel: 'TypeModel',
    query: 'query',
    body: 'data',
    parameters: 'params'
  },
  adapterPath: '../_adapter.ts'
}

export interface GeneratorContext extends ParserContext {
  fs: IFs
  vol: Volume
}

export async function generateFromCtx(ctx: GeneratorContext) {
  const groupedApis = Object.values(groupBy(ctx.apis, (n) => n.path))

  const spinner = createSpinner(
    'Generate code from spec',
    groupedApis.length
  ).start()

  for (const groupedApi of groupedApis) {
    await generateApiByPath(ctx, groupedApi, ctx.fs)
    spinner.plusOne()
  }

  spinner.done()

  generateIndexFiles(ctx.fs)
}

function generateIndexFiles(vfs: IFs) {
  return _generateIndexFiles(vfs, '/')

  function _generateIndexFiles(vfs: IFs, input: string) {
    const files = vfs.readdirSync(input) as string[]

    const hasIndex = files.includes('index.ts')

    if (hasIndex) {
      console.error('Can not create index file, for:', input)
      return
    }

    {
      // generate index file
      const tsFiles = files.filter((n) => n.endsWith('.ts'))
      const theSameNameFolder = files.filter((item) =>
        tsFiles.includes(item + '.ts')
      )

      theSameNameFolder.forEach((name) => {
        const _file = path.join(input, name + '.ts')
        let code = vfs.readFileSync(_file)
        code = [code, `export * from './${name}/index.ts'`].join('\n\n')

        vfs.writeFileSync(_file, code)
      })

      const includeFiles = files.filter((n) => !theSameNameFolder.includes(n))

      const indexFileCodes: string[] = includeFiles.map((item) => {
        const name = camelCase(item.replace(/\.ts$/, ''))

        return `export * as ${name} from './${item}'`
      })

      const codes = [`//@ts-ignore\n//@ts-nocheck`, ...indexFileCodes]

      vfs.writeFileSync(path.join(input, 'index.ts'), codes.join('\n'))
    }

    for (const file of files) {
      const _input = path.join(input, file.toString())
      const isDirectory = vfs.statSync(_input).isDirectory()

      if (isDirectory) {
        _generateIndexFiles(vfs, _input)
      }
    }
  }
}

async function generateApiByPath(
  ctx: ParserContext,
  groupedApi: APIConfig[],
  fs: IFs
) {
  const apiPath = groupedApi.at(0)!.path
  const outputFilePath = normalizeFilePath(apiPath)

  const adapterRelativePath = path
    .join(path.relative(path.dirname(outputFilePath), '/'), config.adapterPath)
    .replaceAll('\\', '/')

  const codes: string[] = [
    `// @ts-ignore\n// @ts-nocheck`,
    `import { ${config.nameMapper.requestAdapterName}, type RequestConfig } from '${adapterRelativePath}'`
  ]

  const types = Object.assign({}, ...groupedApi.map((item) => item.types))

  const schema: SchemaObject = {
    ...ctx.schema,
    type: 'object',
    properties: types,
    required: Object.keys(types)
  }

  const typeCodes = await generateInterface(config.nameMapper.apiModel, schema)
  codes.push(typeCodes)

  const apiCodes = groupedApi
    .map((api) => generateApiMethodCode(api))
    .join('\n\n')
  codes.push(apiCodes)

  fs.mkdirSync(dirname(outputFilePath), { recursive: true })

  fs.writeFileSync(outputFilePath, codes.join('\n\n'))
}

function generateApiMethodCode(api: APIConfig) {
  const comments = [
    '/**',
    `* ${api.summary || ''}`,
    '*',
    `* ${api.description || ''}`,
    '*',
    `* url path: ${api.path}`,
    api.bodyTypeIsFormData &&
      api.bodyType &&
      `* @param data FormData keys: [${Object.keys(
        ('properties' in api.bodyType.schema &&
          api.bodyType.schema.properties) ||
          {}
      ).join(', ')}]`,
    api.tags && `* @tags ${api.tags.join(', ')}`,
    '*/'
  ].filter(Boolean)

  const params = [
    `method: '${api.method}',`,
    `    url: \`${generateRequestUrl(api)}\`,`,
    (api.bodyType || api.bodyTypeIsFormData) &&
      `    body: ${config.nameMapper.body},`,
    api.queryType && `    query: ${config.nameMapper.query},`,
    api.paramsType && `    params: ${config.nameMapper.parameters},`,
    'config,'
  ].filter(Boolean)

  const codes = [
    comments.join('\n'),
    `export const $${api.method.toLowerCase()} = (${generateRequestParameters(
      api
    )}) => {
    return ${config.nameMapper.requestAdapterName}<${generateResponseType(
      api
    )}>({
      ${params.join('\n')}
    })
  }`
  ]

  return codes.join('\n')
}

function generateRequestUrl(api: APIConfig): string {
  const parseParamReg = /\{([^}]+)\}/g

  return api.path.replace(
    parseParamReg,
    `\${${config.nameMapper.parameters}.\$1}`
  )
}

function normalizeFilePath(apiPath: string) {
  let filePath = apiPath
    .split('/')
    .map((item) => (/^\{[^\}]+\}$/.test(item) ? `_${item.slice(1, -1)}` : item))
    .join('/')

  if (filePath === '/') {
    filePath = '/_'
  }

  return filePath + '.ts'
}

interface ParameterConfig {
  name: string
}

export interface ParametersConfig {
  query: ParameterConfig
  body: ParameterConfig
  param: ParameterConfig
}

function generateRequestParameters(api: APIConfig): string {
  const params: string[] = []

  if (api.paramsType) {
    params.push(
      generateParameterType(api.paramsType, {
        name: config.nameMapper.parameters
      })
    )
  }

  if (api.bodyTypeIsFormData) {
    params.push(`${config.nameMapper.body}: FormData`)
  } else if (api.bodyType) {
    params.push(
      generateParameterType(api.bodyType, { name: config.nameMapper.body })
    )
  }

  if (api.queryType) {
    params.push(
      generateParameterType(api.queryType, { name: config.nameMapper.query })
    )
  }

  params.push(`config?: RequestConfig`)
  return params.join(', ')
}

function generateParameterType(
  parameterSchema: APIParameterConfig,
  opt: ParameterConfig
): string {
  if (!parameterSchema) return ''

  const typeName = parameterSchema.name

  let type = `${config.nameMapper.apiModel}["${typeName}"]`

  return `${opt.name}: ${type}`
}

function generateResponseType(api: APIConfig): string {
  if (!api.responseType) return 'void'

  return `${`${config.nameMapper.apiModel}["${api.responseType.name}"]`}`
}
