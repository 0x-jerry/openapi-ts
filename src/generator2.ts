import { groupBy } from 'lodash-es'
import type { ParserContext, APIConfig, APIParameterConfig } from './parser2'
import { type IFs, Volume, createFsFromVolume } from 'memfs'
import path, { dirname } from 'path'
import { type SchemaObject } from 'openapi-typescript'
import { generateInterface } from './generateTypes'

const config = {
  nameMapper: {
    apiModel: 'TypeModel',
    query: 'query',
    body: 'data',
    parameters: 'params',
  },
  adapterPath: '_http.ts',
}

export async function generate(ctx: ParserContext) {
  const groupedApis = Object.values(groupBy(ctx.apis, (n) => n.path))

  const vfs = createFsFromVolume(new Volume())

  const p = groupedApis.map((groupedApi) => generateApiByPath(ctx, groupedApi, vfs))

  await Promise.all(p)

  // todo
  // generate index.ts

  return vfs
}

async function generateApiByPath(ctx: ParserContext, groupedApi: APIConfig[], fs: IFs) {
  let outputFilePath = groupedApi.at(0)!.path
  outputFilePath = normalizeApiPath(outputFilePath) + '.ts'

  const adapterRelativePath = path.join(
    path.relative(path.dirname(outputFilePath), '/'),
    config.adapterPath
  )

  const codes: string[] = [
    `// @ts-ignore\n// @ts-nocheck`,
    `import { _http, type RequestConfig } from '${adapterRelativePath}'`,
  ]

  const types = Object.assign({}, ...groupedApi.map((item) => item.types))

  const schema: SchemaObject = {
    ...ctx.schema,
    type: 'object',
    properties: types,
    required: Object.keys(types),
  }

  const typeCodes = await generateInterface(config.nameMapper.apiModel, schema)
  codes.push(typeCodes)

  const apiCodes = groupedApi.map((api) => generateApi(api)).join('\n\n')
  codes.push(apiCodes)

  fs.mkdirSync(dirname(outputFilePath), { recursive: true })

  fs.writeFileSync(outputFilePath, codes.join('\n\n'))
}

function generateApi(api: APIConfig) {
  const params = [
    `    url: \`${generateRequestUrl(api)}\`,`,
    (api.bodyType || api.bodyTypeIsFormData) && `    body: ${config.nameMapper.body},`,
    api.queryType && `    query: ${config.nameMapper.query},`,
    'config',
  ].filter(Boolean)

  return `export const $${api.method.toLowerCase()} = (${generateRequestParameters(api)}) => {
    return _http.${api.method}<${generateResponseType(api)}>({
      ${params.join('\n')}
    })
  }`
}

function generateRequestUrl(api: APIConfig): string {
  const parseParamReg = /\{([^}]+)\}/g

  return api.path.replace(parseParamReg, `\${${config.nameMapper.parameters}.\$1}`)
}

function normalizeApiPath(path: string) {
  return path
    .split('/')
    .map((item) => (/^\{[^\}]+\}$/.test(item) ? `_${item.slice(1, -1)}` : item))
    .join('/')
}

interface ParameterConfig {
  name: string
}

export interface ParametersConfig {
  query: ParameterConfig
  body: ParameterConfig
  param: ParameterConfig
}

function generateRequestParameters(api: APIConfig): string {
  const params: string[] = []

  if (api.paramsType) {
    params.push(generateParameterType(api.paramsType, { name: config.nameMapper.parameters }))
  }

  if (api.bodyTypeIsFormData) {
    params.push(`${config.nameMapper.body}: FormData`)
  } else if (api.bodyType) {
    params.push(generateParameterType(api.bodyType, { name: config.nameMapper.body }))
  }

  if (api.queryType) {
    params.push(generateParameterType(api.queryType, { name: config.nameMapper.query }))
  }

  params.push(`config?: RequestConfig`)
  return params.join(', ')
}

function generateParameterType(parameterSchema: APIParameterConfig, opt: ParameterConfig): string {
  if (!parameterSchema) return ''

  const typeName = parameterSchema.name

  let type = `${config.nameMapper.apiModel}["${typeName}"]`

  return `${opt.name}: ${type}`
}

function generateResponseType(api: APIConfig): string {
  if (!api.responseType) return 'void'

  return `${`${config.nameMapper.apiModel}["${api.responseType.name}"]`}`
}
