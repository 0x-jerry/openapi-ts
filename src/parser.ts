import type {
  MediaTypeObject,
  OpenAPI3,
  OperationObject,
  ParameterObject,
  PathsObject,
  ReferenceObject,
  RequestBodyObject,
  ResponseObject,
  SchemaObject,
} from 'openapi-typescript'
import { getRef, isRef } from './helper'
import { ensureArray, PascalCase } from '@0x-jerry/utils'
import { unifySchema } from './normalize'
import { convertPathToName } from './utils'

export const METHODS = [
  //
  'get',
  'put',
  'post',
  'delete',
  'patch',
] as const

export type APIMethod = (typeof METHODS)[number]

export type InType = 'query' | 'header' | 'path' | 'cookie' | 'formData' | 'body'

type IJSONSchema = SchemaObject | ReferenceObject

export interface APIParameterConfig {
  name: string
  schema: IJSONSchema
}

export interface APIConfig {
  name: string
  summary?: string

  tags?: string[]

  description?: string
  /**
   * api url path
   */
  path: string

  method: APIMethod

  /**
   * request path parameters
   *
   * ex. `/user/{id}`
   */
  paramsType?: APIParameterConfig

  /**
   * request query parameters
   *
   * ex. `/user?id=?`
   */
  queryType?: APIParameterConfig

  /**
   * request body, only support json format
   */
  bodyType?: APIParameterConfig

  bodyTypeIsFormData: boolean

  responseType?: APIParameterConfig

  /**
   * all output types
   */
  types: Record<string, IJSONSchema>

  /**
   * reference types
   */
  refTypes: Record<string, IJSONSchema>
}

export interface ParserContext {
  /**
   * openapi v3 schema
   */
  schema: OpenAPI3

  /**
   * api config list
   */
  apis: APIConfig[]
}

export interface ParseOpenAPIOption {
  /**
   * use api to convert schema
   */
  forceUseApi?: boolean
}

export async function parseOpenAPI(schema: any, opt?: ParseOpenAPIOption) {
  const ctx: ParserContext = {
    schema: await unifySchema(schema, opt?.forceUseApi),
    apis: [],
  }

  parsePaths(ctx, ctx.schema.paths || {})

  return ctx
}

function parsePaths(ctx: ParserContext, paths: PathsObject) {
  for (const [path, _conf] of Object.entries(paths)) {
    const conf = getRef(ctx, _conf)
    const methods = Object.keys(conf).filter((m: any) => METHODS.includes(m)) as APIMethod[]

    for (const method of methods) {
      const op = getRef(ctx, conf[method])

      if (!op) return

      const api = parseAPI(
        ctx,
        {
          // inherit description
          ...op,
        },
        path,
        method,
      )

      ctx.apis.push(api)
    }
  }
}

function parseAPI(ctx: ParserContext, op: OperationObject, path: string, method: APIMethod) {
  const params = ensureArray(op.parameters || []).map((n) => getRef(ctx, n))

  const api: APIConfig = {
    name: convertPathToName(path),
    summary: op.summary,
    description: op.description,
    tags: op.tags,
    path: path,
    method: method,
    bodyTypeIsFormData: method === 'get' ? false : isFormData(ctx, getRef(ctx, op.requestBody)),
    types: {},
    refTypes: {},
  }

  const paramsTypeSchema = parsePathParams(params, path)
  if (paramsTypeSchema) {
    const name = api.name + PascalCase(api.method) + 'Param'
    addApiTypeDef(api, name, paramsTypeSchema)

    api.paramsType = {
      name,
      schema: paramsTypeSchema,
    }
  }

  const queryTypeSchema = parseQueryParams(params)
  if (queryTypeSchema) {
    const name = api.name + PascalCase(api.method) + 'Query'
    addApiTypeDef(api, name, queryTypeSchema)

    api.queryType = {
      name,
      schema: queryTypeSchema,
    }
  }

  const bodyTypeSchema =
    method === 'get' ? null : parseRequestBodyType(ctx, getRef(ctx, op.requestBody))
  if (bodyTypeSchema) {
    const name = api.name + PascalCase(api.method) + 'RequestBody'
    addApiTypeDef(api, name, bodyTypeSchema)

    api.bodyType = {
      name,
      schema: bodyTypeSchema,
    }
  }

  const responseTypeSchema = parseResponseType(ctx, op.responses)
  if (responseTypeSchema) {
    const name = api.name + PascalCase(api.method) + 'Response'
    addApiTypeDef(api, name, responseTypeSchema)

    api.responseType = {
      name,
      schema: responseTypeSchema,
    }
  }

  // generate refTypes
  api.refTypes = generateRefTypes(ctx, api.types)
  return api
}

function addApiTypeDef(api: APIConfig, name: string, schema: IJSONSchema) {
  api.types[name] = schema
}

function generateRefTypes(ctx: ParserContext, types: APIConfig['types']) {
  const schemas = Object.values(types)
  const result: APIConfig['refTypes'] = {}

  const processed = new Set<IJSONSchema>()

  for (const schema of schemas) {
    searchRefType(ctx, schema)
  }

  return result

  function searchRefType(ctx: ParserContext, schema: IJSONSchema): void {
    if (processed.has(schema)) {
      return
    }

    processed.add(schema)

    if (isRef(schema)) {
      const s = getRef(ctx, schema)

      result[schema.$ref] = s

      searchRefType(ctx, s)
    } else if (schema.type === 'object') {
      const schemas = Object.values(schema.properties || {})

      for (const prop of schemas) {
        searchRefType(ctx, prop)
      }

      if (schema.additionalProperties) {
        searchRefType(ctx, schema.additionalProperties as IJSONSchema)
      }
    } else if (schema.type === 'array') {
      const schemas = ensureArray(schema.items)

      for (const schema of schemas) {
        searchRefType(ctx, schema)
      }
    }
  }
}

function parseResponseType(
  ctx: ParserContext,
  responses?: Record<string, ReferenceObject | ResponseObject>,
) {
  const content = getRef(ctx, responses?.['200'])?.content

  const schema = (content?.['application/json'] || content?.['*/*'])?.schema

  return schema || null
}

function isFormData(ctx: ParserContext, body: RequestBodyObject | undefined): boolean {
  const contentType = getRef(ctx, body?.content)

  if (!contentType) return false

  return (
    !(contentType['*/*'] || contentType['application/json']) && !!contentType['multipart/form-data']
  )
}

function parseRequestBodyType(
  ctx: ParserContext,
  body: RequestBodyObject | undefined,
): IJSONSchema | null {
  const contentType = getRef(ctx, body?.content)

  if (!contentType) return null

  const content = (contentType['*/*'] ||
    contentType['application/json'] ||
    contentType['multipart/form-data']) as MediaTypeObject

  return content?.schema || null
}

function parseQueryParams(parameters: ParameterObject[]): SchemaObject | null {
  const types = parameters.filter((n) => n.in === 'query' && n.schema && n.name)

  if (!types.length) return null

  const schema: SchemaObject = {
    type: 'object',
    properties: {},
  }

  types.forEach((param) => {
    schema.properties![param.name!] = {
      description: param.description,
      ...param.schema!,
    }
  })

  return schema
}

function parsePathParams(params: ParameterObject[], path: string): SchemaObject | null {
  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions/Assertions#other_assertions
  const pathParamsList: string[] = path.match(/(?<=\{)[a-z0-9_]+(?=\})/gi) || []

  if (!pathParamsList.length) return null

  const schema: SchemaObject = {
    type: 'object',
    properties: {},
  }

  const pathParams = params.filter((p) => p.in === 'path' && p.name)

  pathParams.forEach((param) => {
    schema.properties![param.name!] = {
      description: param.description,
      ...param.schema!,
    }

    pathParamsList.splice(pathParamsList.indexOf(param.name!), 1)
  })

  // add missing path parameters
  pathParamsList.forEach((name) => {
    schema.properties![name] = {
      type: 'string',
    }
  })

  schema.required = Object.keys(schema.properties!)

  return schema
}
