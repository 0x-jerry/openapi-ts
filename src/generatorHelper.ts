import fsp from 'fs-extra'
import { parseOpenAPI, type APIConfig } from './parser'
import { createFsFromVolume, Volume, type IFs } from 'memfs'
import path from 'node:path'
import { generateFromCtx, type GenerateOption, type GeneratorContext } from './generator'

export interface GenerateClientCodesOptions {
  // biome-ignore lint/suspicious/noExplicitAny: <explanation>
  schema: any

  /**
   * @default 'nested'
   */
  apiStyle?: GenerateOption['style']

  /**
   * @default 'api/generated'
   */
  output?: string

  /**
   * Filter api
   * @param api
   */
  filter?: (api: APIConfig) => boolean

  /**
   * Format output code.
   */
  format?: boolean

  /**
   * Clean output dir before write file.
   */
  clean?: boolean
}

async function generateClientCodes(opt: GenerateClientCodesOptions) {
  const parser = await parseOpenAPI(opt.schema)

  if (opt.filter) {
    parser.apis = parser.apis.filter(opt.filter)
  }

  const vol = new Volume()
  const ctx: GeneratorContext = {
    ...parser,
    vol,
    fs: createFsFromVolume(vol),
  }

  await generateFromCtx(ctx, {
    style: opt.apiStyle || 'nested',
  })

  if (opt.format) {
    await formatCodes(ctx.fs)
  }

  return ctx
}

export async function generate(opt: GenerateClientCodesOptions) {
  const option: GenerateClientCodesOptions = Object.assign(
    {
      output: 'api/generated',
      format: false,
      clean: false,
    },
    opt,
  )

  const ctx = await generateClientCodes(option)

  if (option.output) {
    await writeToDisk(ctx.fs, option.output, {
      clean: opt.clean,
    })
  }
}

async function formatCodes(vfs: IFs, dir = '/') {
  const files = vfs.readdirSync(dir)

  for (const file of files) {
    const filePath = path.join(dir, file.toString())
    const isDirectory = vfs.statSync(filePath).isDirectory()

    if (isDirectory) {
      await formatCodes(vfs, filePath)
    } else if (filePath.endsWith('.ts')) {
      const prettier = await import('prettier')

      const content = vfs.readFileSync(filePath)

      try {
        const formattedContent = await prettier.format(content.toString(), {
          parser: 'typescript',
        })
        vfs.writeFileSync(filePath, formattedContent)
      } catch (error) {
        console.warn('Format code failed:', filePath)
      }
    }
  }
}

async function writeToDisk(vfs: IFs, output: string, opt: { clean?: boolean } = {}) {
  const out = path.resolve(output)

  if (opt.clean && (await fsp.pathExists(out))) {
    await fsp.emptyDir(out)
  }

  return _writeToDisk(vfs, '/', out)

  async function _writeToDisk(vfs: IFs, input: string, output: string) {
    const files = vfs.readdirSync(input)

    for (const file of files) {
      const _input = path.join(input, file.toString())
      const _out = path.join(output, file.toString())

      const isDirectory = vfs.statSync(_input).isDirectory()

      if (isDirectory) {
        _writeToDisk(vfs, _input, _out)
      } else {
        // write to disk
        await fsp.ensureDir(output)

        await fsp.writeFile(_out, vfs.readFileSync(_input))
      }
    }
  }
}
