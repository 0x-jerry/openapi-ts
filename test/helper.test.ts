import { readFile } from 'fs/promises'
import path from 'path'
import { replaceSchemaType } from '../src/helper'
import { parseOpenAPI } from '../src/parser'

describe('helper', () => {
  it('replaceSchemaType', async () => {
    const content = await readFile(path.join(import.meta.dirname, 'schema/v3.json'), 'utf8')

    const schema = JSON.parse(content)
    const ctx = await parseOpenAPI(schema)

    replaceSchemaType(ctx, {
      type: 'string',
      targetType: 'number',
    })

    expect(ctx.schema.components!.schemas!.Employee).toMatchSnapshot()
  })
})
