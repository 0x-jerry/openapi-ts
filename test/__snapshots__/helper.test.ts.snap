// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`helper > replaceSchemaType 1`] = `
{
  "additionalProperties": false,
  "properties": {
    "id": {
      "format": undefined,
      "nullable": true,
      "readOnly": true,
      "type": "number",
    },
    "name": {
      "format": undefined,
      "nullable": true,
      "readOnly": true,
      "type": "number",
    },
  },
  "type": "object",
}
`;
